sql/hive-thriftserver/src/main/scala/org/apache/spark/sql/hive/thriftserver/SparkSQLSessionManager.scala

```diff
@@ -75,6 +75,9 @@ private[hive] class SparkSQLSessionManager(hiveServer: HiveServer2, sqlContext:
 
   override def closeSession(sessionHandle: SessionHandle): Unit = {
     HiveThriftServer2.listener.onSessionClosed(sessionHandle.getSessionId.toString)
+    val ctx = sparkSqlOperationManager.sessionToContexts.getOrDefault(sessionHandle, sqlContext)
+    ctx.sparkSession.sessionState.catalog.getTempViewNames().foreach(ctx.uncacheTable)
```

Comment:
Oh, I meant call uncacheTable inside clearTempTables. Is it possible?

Reply:
I think it can not call `uncacheTable ` inside `clearTempTables`. There is no sqlContext reference in `SessionCatalog`

Reply:
Ah right. The `clearTempTables()` call doesn't really matter then. Maybe revert that part just to keep the change simple.

It seems like a good thing to fix and this is straightforward, I just keep wondering if there's a slightly better place to manage it, or whether there are more cases like this that need to be addressed. Eh, @cloud-fan do you have an opinion or is this a fine way to do it?

Reply:
reverted

Reply:
shall we do it in `SparkSesson.stop`? The code path here seems thrift server specific.

Reply:
`SparkSesson.stop` will stop underlying `SparkContext`. I don't think it's the right place.
https://github.com/apache/spark/blob/882f54b0a323fb5cd827d600b3c3332e1fcdf65a/sql/core/src/main/scala/org/apache/spark/sql/SparkSession.scala#L672

Reply:
What's more, SQL session close in thrift server doesn't reach `SparkSesson.stop`.

Reply:
hmm the behavior is really weird. I think we should track all the active sessions and only stop the spark context if the session being closed is the last one.

Reply:
I can work that in another ticket. In current `SparkSQLSessionManager` implementation, each query will create a new SparkSession and store in the map `SparkSQLOperationManager.sessionToContexts`. When a session close, this session will be removed from this map and GC later. `SparkSesson.stop` won't be invoked at all.

Reply:
@cloud-fan is that a somewhat different point? I'm not an expert here. Seems like this is about unpersisting cached temp views in the session when the session is done, not stopping Spark when all active sessions are closed. As in, is it valid to re-start a new session in the same context?

Reply:
Ideally we should uncache the temp views when the session is closed, but seems we don't have such an action called "close session" in Spark SQL. `SparkSession.close` closes the SparkContext and thus closes all sessions.

If we have "close session" action in thrift server, I think it's corrected to put the logic here.

Reply:
Thanks @cloud-fan and @srowen . So the patch is correct so far, right?

Reply:
Since there is no place to trace all live spark sessions in a non-thrift-server application. For example spark session can be stored in thread local variable and be garbage collection when user's thread exit. So it still has memory leak problem in an application. I think I can open a new ticket to manage/trace all created spark session. For now, could we accept this due to thrift server is official implemented multiple session application.

Reply:
https://issues.apache.org/jira/browse/SPARK-29972